jbellis/jamm
# Improve performance by 10x

This PR improves jamm's performance to measure an arbitrary object graph by factor 10, proven by a microbenchmark.

To achieve this, a bunch of changes have been made:
* Remove unused code and untested code paths
* Separate jamm configuration and runtime (breaking API change) - i.e. introduce a builder-pattern and distinct implementations for the "specification", "instrumentation" and "unsafe" modes.
* Bunch of optimizations like
  * a specialized "visited-objects-set" for jamm's `measureDeep()`
  * specialized/optimal code paths for different jamm-modes (was using a lot of `if`s in the hot-paths before)
  * memoizing calculated values (fields of a class like #39, class-inclusion-checks
  * added new modes to measure `ByteBuffer` for Apache Cassandra (see below)
  * deprecate jamm's "unsafe" + "specification" modes to make people aware that using `sun.misc.Unsafe` is wrong and "specification" (actually: just guessing) cannot be aware of all current and especially future object layouts provided by JDKs/JVMs
  * specialized `sizeOfArray()` functions
* A bunch of `MemoryMeter.Guess` enum values have been deprecated - new ones `ALWAYS_INSTRUMENTATION` + `BEST` introduced
* With [JDK-8249196](https://bugs.openjdk.java.net/browse/JDK-8249196) in mind, these functionalities have been removed:
  * ability to debug (dump the object graph traversal) - this also affected production performance
  * ability to count the number of objects in a graph - this was untested
  * allow easy integration of a new jamm-mode "runtime" for JEP/JDK-8249196
  * `@Unmetered` no longer allowed on fields, only on types (classes + interfaces)

This change works w/ C*. A preliminary draft for the C* changes is in the last commit [of this branch](https://github.com/snazy/cassandra/commits/jamm-runtime-deepsizeof)

This work is separated into 4 commits:
* Rename readme to `.md`
* Add microbenchmark (jmh)
* Switch to Gradle to avoid jmh-dependencies leaking into the published pom and not refactor jamm to a multi-module maven-project
* Refactoring of jamm to gain a 10x improvement (benchmark results in the readme)

Obsoletes #40 

@szegedi maybe this is interesting for you?
